src/ts_catalog/dimension_partition.c

```diff
@@ -0,0 +1,432 @@
+		replica_nodes = lappend(replica_nodes, list_nth(data_nodes, list_index));
+	}
+
+	return replica_nodes;
+}
+
+static HeapTuple
+create_dimension_partition_tuple(Relation rel, const DimensionPartition *dp)
+{
+	TupleDesc tupdesc = RelationGetDescr(rel);
+	Datum values[Natts_dimension_partition];
+	bool nulls[Natts_dimension_partition] = { false };
+	int i = 0;
+
+	values[AttrNumberGetAttrOffset(Anum_dimension_partition_dimension_id)] =
+		Int32GetDatum(dp->dimension_id);
+	values[AttrNumberGetAttrOffset(Anum_dimension_partition_range_start)] =
+		Int64GetDatum(dp->range_start);
+
+	if (dp->data_nodes == NIL)
+	{
+		nulls[AttrNumberGetAttrOffset(Anum_dimension_partition_data_nodes)] = true;
+	}
+	else
+	{
+		Datum *dn_datums = palloc(sizeof(Datum) * list_length(dp->data_nodes));
+		ArrayType *dn_arr;
+		ListCell *lc;
+
+		foreach (lc, dp->data_nodes)
+		{
+			const char *dn = lfirst(lc);
+			dn_datums[i++] = CStringGetDatum(dn);
+		}
+
+		dn_arr = construct_array(dn_datums,
+								 list_length(dp->data_nodes),
+								 NAMEOID,
+								 NAMEDATALEN,
+								 false,
+								 TYPALIGN_CHAR);
+		values[AttrNumberGetAttrOffset(Anum_dimension_partition_data_nodes)] =
+			PointerGetDatum(dn_arr);
+	}
+
+	return heap_form_tuple(tupdesc, values, nulls);
+}
+
+static void
+dimension_partition_info_delete(int dimension_id, int scanflags, LOCKMODE lockmode)
+{
+	ScanIterator it;
+	CatalogSecurityContext sec_ctx;
+
+	it = ts_dimension_partition_scan_iterator_create(lockmode);
+	ts_dimension_partition_scan_iterator_set_dimension_id(&it, dimension_id, NULL);
+	it.ctx.flags = scanflags;
+
+	ts_catalog_database_info_become_owner(ts_catalog_database_info_get(), &sec_ctx);
+
+	ts_scanner_foreach(&it)
+	{
+		const TupleInfo *ti = ts_scan_iterator_tuple_info(&it);
+		ts_catalog_delete_tid_only(ti->scanrel, &ti->slot->tts_tid);
+	}
+
+	ts_catalog_restore_user(&sec_ctx);
+	ts_scan_iterator_close(&it);
+}
+
+/*
+ * Delete all dimension partitions for a given dimension.
+ */
+void
+ts_dimension_partition_info_delete(int dimension_id)
+{
+	dimension_partition_info_delete(dimension_id, SCANNER_F_NOFLAGS, RowExclusiveLock);
+}
+
+DimensionPartitionInfo *
+ts_dimension_partition_info_recreate(int32 dimension_id, unsigned int num_partitions,
+									 List *data_nodes, int replication_factor)
+{
+	int64 partition_size = DIMENSION_SLICE_CLOSED_MAX / ((int64) num_partitions);
+	int64 range_start = DIMENSION_SLICE_MINVALUE;
+	Catalog *catalog = ts_catalog_get();
+	Oid relid = catalog_get_table_id(catalog, DIMENSION_PARTITION);
+	DimensionPartitionInfo *dpi;
+	RBTree *tree = NULL;
+	Relation rel;
+	unsigned int i;
+
+	Assert(num_partitions > 0);
+	Assert(data_nodes == NIL || replication_factor > 0);
+
+	/* Delete all partitions for the dimension */
+	dimension_partition_info_delete(dimension_id, SCANNER_F_KEEPLOCK, RowExclusiveLock);
+
+	/* Lock already held */
+	rel = table_open(relid, NoLock);
+
+	tree = rbt_create(sizeof(DimensionPartition),
```

Comment:
From where are we getting this rbt_create API?

Reply:
It's standard PG library

Reply:
Switched to array+binary search, so no longer using rb-tree.